### PR TITLE
Charlesmchen/unencrypted headers

### PR DESCRIPTION
**Pull request by charlesmchen**
PTAL @michaelkirk

I intend to open two separate PRs: One against our fork (and against the commit that corresponds to the latest Cocoapods release of YapDatabase), and a separate PR against upstream (and against their latest master).

### CODE REVIEW

File: YapDatabase/Utilities/YapDatabaseCryptoUtils.h [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+extern const NSUInteger kSqliteHeaderLength;
+extern const NSUInteger kSQLCipherSaltLength;
+extern const NSUInteger kSQLCipherDerivedKeyLength;
+extern const NSUInteger kSQLCipherKeySpecLength;
+
+typedef void (^YapDatabaseSaltBlock)(NSData *saltData);
+typedef void (^YapDatabaseKeySpecBlock)(NSData *keySpecData);
+
+// This class contains utility methods for use with SQLCipher encrypted
```
Comment (charlesmchen) [inline YapDatabase/Utilities/YapDatabaseCryptoUtils.h]: I ended up making this the main comment.

**Comment by charlesmchen** [comment]
See: https://github.com/yapstudios/YapDatabase/issues/436

**Comment by charlesmchen** [comment]
@michaelkirk I've merged this into a new branch `release/unencryptedHeaders` in this fork so that historical references to this commit are valid when building legacy versions of Signal.

**Comment by charlesmchen** [comment]
See: https://github.com/sqlcipher/sqlcipher/issues/255

**Review by michaelkirk — COMMENTED** [review]
No serious concerns with the logic, but some doc suggestions.

I'm eager to get this in front of the Yap so we can get this upstreamed. They may not be excited about all the helper methods cribbed from our codebase (YapFail, etc.), but I guess we'll see.

File: YapDatabase/Utilities/YapDatabaseCryptoUtils.h [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+// Solution
+//
+// The solution is to have SQLCipher encrypt everything _except_ the first 32 bytes of
+// the Sqlite file, which corresponds to the first part of the Sqlite header.  This is
+// accomplished using the cipher_plaintext_header_size PRAGMA.
+//
+// The header does not contain any user data.  See:
+// https://www.sqlite.org/fileformat.html#the_database_header
+//
+// However, Sqlite normally uses the first 16 bytes of the Sqlite header to store
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply YapDatabase/Utilities/YapDatabaseCryptoUtils.h]: Good catch.

File: YapDatabase/Utilities/YapDatabaseCryptoUtils.h [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+// a salt value.  Therefore when using unencrypted headers, it is also necessary
+// to explicitly specify a salt value.
+//
+// It is possible to convert SQLCipher databases with encrypted headers to use
+// unencrypted headers.  However, during this conversion, the salt must be extracted
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply YapDatabase/Utilities/YapDatabaseCryptoUtils.h]: Done.

File: YapDatabase/Utilities/YapDatabaseCryptoUtils.h [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+//
+// * Call the convertDatabaseIfNecessary method of this class _before_
+//   trying to open any YapDatabase that may need to be converted.
+// * This method will have no effect if the YapDatabase has already been converted.
+// * This method should always be pretty fast, and should be safe to
+//   call from within [UIApplicationDelegate application: didFinishLaunchingWithOptions:].
+// * If convertDatabaseIfNecessary converts the database, it will use its
+//   saltBlock and keySpecBlock parameters to inform you of the salt
+//   and keyspec for this database.  These values will be needed when
+//   opening the database, so they should presumably stored in the
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply YapDatabase/Utilities/YapDatabaseCryptoUtils.h]: Done.

File: YapDatabase/Utilities/YapDatabaseCryptoUtils.m [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+}
+
++ (nullable NSData *)deriveDatabaseKeyForPassword:(NSData *)passwordData saltData:(NSData *)saltData
+{
+    YapAssert(passwordData.length > 0);
+    YapAssert(saltData.length == kSQLCipherSaltLength);
+
+    unsigned char *derivedKeyBytes = malloc((size_t)kSQLCipherDerivedKeyLength);
+    YapAssert(derivedKeyBytes);
+    // See: PBKDF2_ITER.
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply YapDatabase/Utilities/YapDatabaseCryptoUtils.m]: Done.

File: YapDatabase/YapDatabase.m [hunk 6]
```diff
@@ -820,16 +820,27 @@ - (BOOL)configureDatabase:(BOOL)isNewDatabaseFile
...
-		
+        NSData *_Nullable keyData = nil;
+        if (options.cipherKeySpecBlock)
+        {
+            keyData = options.cipherKeySpecBlock();
+            if (!keyData)
+            {
+                NSAssert(NO, @"YapDatabaseOptions.cipherKeySpecBlock cannot return nil!");
+                return NO;
+            }
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply YapDatabase/YapDatabase.m]: > If the keySpecBlock produces something of the wrong length, it'll be run through the kdf anyway.

Is that true?  Key specs are passed using a different format than "passwords/keys".  Anyway, added a check around length here.

File: YapDatabase/YapDatabase.m [hunk 7]
```diff
@@ -863,16 +874,107 @@ - (BOOL)configureEncryptionForDatabase:(sqlite3 *)sqlite
...
-			YDBLogError(@"Error setting SQLCipher key: %d %s", status, sqlite3_errmsg(sqlite));
-			return NO;
-		}
+        if (options.cipherKeySpecBlock) {
+            // Use a raw key spec, where the 96 hexadecimal digits are provided
+            // (i.e. 64 hex for the 256 bit key, followed by 32 hex for the 128 bit salt)
+            // using explicit BLOB syntax, e.g.:
+            //
+            // x'98483C6EB40B6C31A448C22A66DED3B5E5E8D5119CAC8327B655C8B5C483648101010101010101010101010101010101'
+            NSString *keySpecString = [NSString stringWithFormat:@"x'%@'", [self hexadecimalStringForData:keyData]];
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply YapDatabase/YapDatabase.m]: Nice simplification.

File: YapDatabase/YapDatabaseOptions.h [hunk 8]
```diff
@@ -223,6 +223,57 @@ typedef NSData *_Nonnull (^YapDatabaseCipherKeyBlock)(void);
...
+ * in your Podfile for this option to take effect.
+ *
+ * cipherSaltBlock will be ignored if you don't also set cipherKeyBlock and cipherUnencryptedHeaderLength.
+ *
+ * For more information, see comments in YapDatabaseCryptoUtils.h.
+ **/
+@property (nonatomic, copy, readwrite) YapDatabaseCipherKeyBlock cipherSaltBlock;
+
+/**
+ * Set a block here that returns the key spec (not the key) for the SQLCipher database.
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply YapDatabase/YapDatabaseOptions.h]: > We should specify the format constraints of the keyspec block. 

Done.

> I'm sort of inclined to explain a bit about when you'd want to use it

Done.

> If you are using cipherKeySpecBlock, don't use cipherKeyBlock.

Done, and for `cipherSaltBlock` too.

> If you are using cipherUnencryptedHeaderLength, you must either use a cipherKeySpecBlock or if you're using a cipherKeyBlock, you must also specify a cipherSaltBlock.

Done.

File: YapDatabase/YapDatabase.m (hunk 6 above)
[the remark this answers is not in the document]
Comment (michaelkirk) [reply YapDatabase/YapDatabase.m]: > Is that true? 

Take a look:

https://github.com/sqlcipher/sqlcipher/blob/c6f709fca81c910ba133aaf6330c28e01ccfe5f8/src/crypto_impl.c#L992

SQLCipher checks for an exact length and verifies the first 2 chars are `x'`

**Comment by GlacierSecurityInc** [comment]
FYI, for anyone struggling with this:

We are currently attempting to move our existing database to shared space and then use convertDatabaseIfNecessary. It seems that if the database was originally created with YapDatabaseCipherCompatability_Version3, the conversion will fail at the point of trying to execute: "PRAGMA journal_mode = WAL." Or at least that's what was happening for us. We needed to run the "PRAGMA cipher_compatibility = YapDatabaseCipherCompatability_Version3" command first, which is not currently an option in the conversion routine. 

If I have time after the overhaul we are doing, I'll try to figure out the most appropriate way to handle this.

**Pull request merged**